### PR TITLE
[SPARK-43920][SQL][CONNECT] Create sql/api module

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

We need a sql/api module to host public API like DataType, Row, etc. This module can be shared between Catalyst and Spark Connect client so that client do not need to depend on Catalyst anymore.

### Why are the changes needed?

Towards Spark Connect client do not need to depend on Catalyst anymore.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

N/A

### CODE REVIEW

**Comment by amaliujia** [comment]
@hvanhovell @cloud-fan

File: sql/api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    <artifactId>spark-sql-api_2.13</artifactId>
+    <packaging>jar</packaging>
+    <name>Spark Project SQL API</name>
+    <url>https://spark.apache.org/</url>
+    <properties>
+        <sbt.project.name>sql-api</sbt.project.name>
+    </properties>
+
+    <dependencies>
```
Comment (hvanhovell) [inline sql/api/pom.xml:37]: Should we add common/util as a dependency?

**Review by hvanhovell — APPROVED** [review]
LGTM

File: sql/api/pom.xml (hunk 1 above)
Comment (amaliujia) [reply sql/api/pom.xml:37]: I will add it once we move code into this module.

**Review by LuciferYang — CHANGES_REQUESTED** [review]
`sqlApi` module shouldn't participate in mima check in the current version

**Comment by amaliujia** [comment]
@LuciferYang right. I updated the SparkBuild to exclude it.

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
Waiting CI

**Comment by hvanhovell** [comment]
Merging.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM.

**Review by dongjoon-hyun — COMMENTED** [review]
BTW, sorry for asking this question after merging.

1. What is the minimal target for migration in order to help `connect` module? According to the PR description, `Row` is also mentioned. So,`sql/catalyst/src/main/scala/org/apache/spark/sql/types` is not enough?
2. Does it cause many MIMA exceptions against old `catalyst` module?

**Comment by amaliujia** [comment]
> BTW, sorry for asking this question after merging.
> 
> 1. What is the minimal target for migration in order to help `connect` module? According to the PR description, `Row` is also mentioned. So,`sql/catalyst/src/main/scala/org/apache/spark/sql/types` is not enough?
> 2. Does it cause many MIMA exceptions against old `catalyst` module?

In current scope, Row, coders, data types are required. However we may discover more during the process.

**Comment by dongjoon-hyun** [comment]
Got it. Let's see it in the actual PR~

**Pull request closed without merging**